library/module_utils/oneview.py

```diff
@@ -344,8 +344,233 @@ class OneViewModuleBase(object):
 
     ONEVIEW_VALIDATE_ETAG_ARGS = dict(validate_etag=dict(type='bool', default=True))
 
+    def __init__(self, additional_arg_spec=None, validate_etag_support=False):
+        """
+        OneViewModuleBase constructor.
+
+        :arg dict additional_arg_spec: Additional argument spec definition.
+        :arg bool validate_etag_support: Enables support to eTag validation.
+        """
+        argument_spec = self._build_argument_spec(additional_arg_spec, validate_etag_support)
+
+        self.module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=False)
+
+        self.resource_client = None
+        self.current_resource = None
+
+        self.state = self.module.params.get('state')
+        self.data = self.module.params.get('data')
+
+        self._check_hpe_oneview_sdk()
+        self._create_oneview_client()
+
+        # Preload params for get_all - used by facts
+        self.facts_params = self.module.params.get('params') or {}
+
+        # Preload options as dict - used by facts
+        self.options = transform_list_to_dict(self.module.params.get('options'))
+
+        self.validate_etag_support = validate_etag_support
+
+    def _build_argument_spec(self, additional_arg_spec, validate_etag_support):
+
+        merged_arg_spec = dict()
+        merged_arg_spec.update(self.ONEVIEW_COMMON_ARGS)
+
+        if validate_etag_support:
+            merged_arg_spec.update(self.ONEVIEW_VALIDATE_ETAG_ARGS)
+
+        if additional_arg_spec:
+            merged_arg_spec.update(additional_arg_spec)
+
+        return merged_arg_spec
+
+    def _check_hpe_oneview_sdk(self):
+        if not HAS_HPE_ONEVIEW:
+            self.module.fail_json(msg=self.HPE_ONEVIEW_SDK_REQUIRED)
+
+    def _create_oneview_client(self):
+        if self.module.params.get('hostname'):
+            config = dict(ip=self.module.params['hostname'],
+                          credentials=dict(userName=self.module.params['username'], password=self.module.params['password'],
+                                           authLoginDomain=self.module.params.get('auth_login_domain', '')),
+                          api_version=self.module.params['api_version'],
+                          image_streamer_ip=self.module.params['image_streamer_hostname'])
+            self.oneview_client = OneViewClient(config)
+        elif not self.module.params['config']:
+            self.oneview_client = OneViewClient.from_environment_variables()
+        else:
+            self.oneview_client = OneViewClient.from_json_file(self.module.params['config'])
+
+    def set_resource_object(self, resource_client):
+        self.resource_client = resource_client
+        name = None
+
+        if self.data and self.data.get("name"):
+            name = self.data["name"]
```

Comment:
Can we do , self.data.get("name", None)  here ?

Reply:
We cannot do this here because 'self.data' will not be dictionary always. 
self.data will be None when the module param doesn't have a data field in it. most of the facts will not be having data field in the params.